apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TextTracestateAppender.java

```diff
@@ -0,0 +1,71 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.impl.transaction;
+
+import javax.annotation.Nullable;
+
+class TextTracestateAppender implements HeaderGetter.HeaderConsumer<String, TraceContext> {
+
+    private static TextTracestateAppender INSTANCE = new TextTracestateAppender();
+
+    static TextTracestateAppender instance() {
+        return INSTANCE;
+    }
+
+    TextTracestateAppender() {
+    }
+
+    @Override
+    public void accept(@Nullable String headerValue, TraceContext traceContext) {
+        if (headerValue == null) {
+            return;
+        }
+        // This means that the tracestate buffer will be allocated from pool only if tracestate headers exist
+        StringBuilder tracestateBuffer = traceContext.getTracestateBuffer();
```

Comment:
I think it's going to be more efficient to just store all headers in a `new ArrayList(2)`. Most of the time, there will only be and more than two seems rather unlikely. Just storing them in a list seems more efficient than pooling a `StringBuilder`, especially because when setting the header, there needs to be an allocation anyways (`StringBuilder#toString`). I'd do the truncation lazily when setting and use a thread-local `StringBuilder` for that.
This optimizes for the likely case that there's only one header that does not exceed the size limit.